Tell LLVM about read-only captures

nikic opened the pull request:
`&Freeze` parameters are not only `readonly` within the function, but any captures of the pointer can also only be used for reads. This can now be encoded using the `captures(address, read_provenance)` attribute.

nikic committed:
Tell LLVM about read-only captures

`&Freeze` parameters are not only `readonly` within the function,
but any captures of the pointer can also only be used for reads.
This can now be encoded using the `captures(address, read_provenance)`
attribute.

rust-timer commented:
Finished benchmarking commit (6bc37934c6eaa84bfd300c54189bec62e15d42ed): [comparison URL](https://perf.rust-lang.org/compare.html?start=577166503aee7290e09374da21f4045c455acfd5&end=6bc37934c6eaa84bfd300c54189bec62e15d42ed&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Benchmarking this pull request means it may be perf-sensitive – we'll automatically label it not fit for rolling up. You can override this, but we strongly advise not to, due to possible changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please do so in sufficient writing along with `@rustbot label: +perf-regression-triaged`. If not, please fix the regressions and do another perf run. If its results are neutral or positive, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=577166503aee7290e09374da21f4045c455acfd5&end=6bc37934c6eaa84bfd300c54189bec62e15d42ed&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.3%, 0.4%]   | 5     |
| Regressions ❌ <br /> (secondary)  | 1.0%  | [0.9%, 1.6%]   | 7     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.8% | [-0.8%, -0.7%] | 5     |
| All ❌✅ (primary)                 | 0.3%  | [0.3%, 0.4%]   | 5     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=577166503aee7290e09374da21f4045c455acfd5&end=6bc37934c6eaa84bfd300c54189bec62e15d42ed&stat=max-rss)
<details>
<summary>Results (secondary 3.9%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 3.9% | [3.9%, 3.9%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=577166503aee7290e09374da21f4045c455acfd5&end=6bc37934c6eaa84bfd300c54189bec62e15d42ed&stat=cycles:u)
<details>
<summary>Results (secondary -2.1%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.1% | [-2.1%, -2.1%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=577166503aee7290e09374da21f4045c455acfd5&end=6bc37934c6eaa84bfd300c54189bec62e15d42ed&stat=size:linked_artifact)
<details>
<summary>Results (secondary 0.0%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.0% | [0.0%, 0.0%] | 23    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

**Bootstrap**: 465.465s -> 464.036s (-0.31%)
**Artifact size**: 377.33 MiB -> 377.30 MiB (-0.01%)

nikic committed:
Adjust test to still show miscompile

The capture of i in assert_ne!() is now known read-only, which
enables early SROA. Block this by passing i to println, where
we currently cannot recognize this.

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17067068913/job/48386707491) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/48386707491) [(plain)](https://github.com/rust-lang/rust/commit/a68e8bee7efaf1d0000a72fe4e348fd9834a419e/checks/48386707491/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [codegen] tests/codegen-llvm/function-arguments.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-19/bin/FileCheck" "--input-file" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/function-arguments/function-arguments.ll" "/checkout/tests/codegen-llvm/function-arguments.rs" "--check-prefix=CHECK" "--allow-unused-prefixes" "--dump-input-context" "100"
stdout: none
--- stderr -------------------------------
/checkout/tests/codegen-llvm/function-arguments.rs:83:11: error: CHECK: expected string not found in input
// CHECK: @readonly_borrow(ptr noalias noundef readonly align 4{{( captures\(address, read_provenance\))}} dereferenceable(4) %_1)
          ^
/checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/function-arguments/function-arguments.ll:320:55: note: scanning from here
define noundef i64 @option_nonzero_int(i64 noundef %x) unnamed_addr #1 {
                                                      ^

Input file: /checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/function-arguments/function-arguments.ll
Check file: /checkout/tests/codegen-llvm/function-arguments.rs

-dump-input=help explains the following input dump.

Input was:
<<<<<<
          .
          .
          .
        220:  %ptr.1 = load ptr, ptr %2, align 8, !nonnull !2, !align !4, !noundef !2 
        221:  %3 = icmp ne ptr %ptr.0, null 
        222:  call void @llvm.assume(i1 %3) 
        223:  %4 = icmp ne ptr %ptr.0, null 
        224:  call void @llvm.assume(i1 %4) 
        225:  call void @llvm.lifetime.start.p0(i64 8, ptr %1) 
        226:  %5 = getelementptr inbounds i8, ptr %ptr.1, i64 8 
        227:  %6 = load i64, ptr %5, align 8, !range !6, !invariant.load !2 
        228:  %7 = getelementptr inbounds i8, ptr %ptr.1, i64 16 
        229:  %8 = load i64, ptr %7, align 8, !range !7, !invariant.load !2 
        230:  store i64 %6, ptr %1, align 8 
        231:  %size = load i64, ptr %1, align 8, !noundef !2 
        232:  call void @llvm.lifetime.end.p0(i64 8, ptr %1) 
        233:  call void @llvm.lifetime.start.p0(i64 8, ptr %0) 
        234:  %9 = getelementptr inbounds i8, ptr %ptr.1, i64 8 
        235:  %10 = load i64, ptr %9, align 8, !range !6, !invariant.load !2 
        236:  %11 = getelementptr inbounds i8, ptr %ptr.1, i64 16 
        237:  %12 = load i64, ptr %11, align 8, !range !7, !invariant.load !2 
        238:  store i64 %12, ptr %0, align 8 
        239:  %align = load i64, ptr %0, align 8, !noundef !2 
        240:  call void @llvm.lifetime.end.p0(i64 8, ptr %0) 
        241:  br label %bb7 
        242:  
        243: bb7: ; preds = %start 
        244:  %13 = sub i64 %align, 1 
        245:  %14 = icmp ule i64 %13, 9223372036854775807 
        246:  call void @llvm.assume(i1 %14) 
        247:  %15 = getelementptr inbounds i8, ptr %layout, i64 8 
        248:  store i64 %size, ptr %15, align 8 
        249:  store i64 %align, ptr %layout, align 8 
        250:  %16 = icmp eq i64 %size, 0 
        251:  br i1 %16, label %bb3, label %bb1 
        252:  
        253: bb3: ; preds = %bb1, %bb7 
        254:  ret void 
        255:  
        256: bb1: ; preds = %bb7 
        257:  %_7 = getelementptr inbounds i8, ptr %self, i64 16 
        258:  %17 = icmp ne ptr %ptr.0, null 
        259:  call void @llvm.assume(i1 %17) 
        260:  %18 = load i64, ptr %layout, align 8, !range !5, !noundef !2 
        261:  %19 = getelementptr inbounds i8, ptr %layout, i64 8 
        262:  %20 = load i64, ptr %19, align 8, !noundef !2 
        263: ; call <alloc::alloc::Global as core::alloc::Allocator>::deallocate 
        264:  call void @"_ZN63_$LT$alloc..alloc..Global$u20$as$u20$core..alloc..Allocator$GT$10deallocate17he6fcd87b9c0c3dc6E"(ptr noalias noundef nonnull readonly align 1 %_7, ptr noundef nonnull %ptr.0, i64 noundef %18, i64 noundef %20) 
        265:  br label %bb3 
        266:  
        267: bb6: ; No predecessors! 
        268:  unreachable 
        269: } 
        270:  
        271: ; Function Attrs: uwtable 
        272: define noundef zeroext i1 @boolean(i1 noundef zeroext %x) unnamed_addr #1 { 
        273: start: 
        274:  ret i1 %x 
        275: } 
        276:  
        277: ; Function Attrs: uwtable 
        278: define i8 @maybeuninit_boolean(i8 %x) unnamed_addr #1 { 
        279: start: 
        280:  ret i8 %x 
        281: } 
        282:  
        283: ; Function Attrs: uwtable 
        284: define noundef zeroext i1 @enum_bool(i1 noundef zeroext %x) unnamed_addr #1 { 
        285: start: 
        286:  ret i1 %x 
        287: } 
        288:  
        289: ; Function Attrs: uwtable 
        290: define i8 @maybeuninit_enum_bool(i8 %x) unnamed_addr #1 { 
        291: start: 
        292:  ret i8 %x 
        293: } 
        294:  
        295: ; Function Attrs: uwtable 
        296: define noundef range(i32 0, 1114112) i32 @char(i32 noundef range(i32 0, 1114112) %x) unnamed_addr #1 { 
        297: start: 
        298:  ret i32 %x 
        299: } 
        300:  
        301: ; Function Attrs: uwtable 
        302: define i32 @maybeuninit_char(i32 %x) unnamed_addr #1 { 
        303: start: 
        304:  ret i32 %x 
        305: } 
        306:  
        307: ; Function Attrs: uwtable 
        308: define noundef i64 @int(i64 noundef %x) unnamed_addr #1 { 
        309: start: 
        310:  ret i64 %x 
        311: } 
        312:  
        313: ; Function Attrs: uwtable 
        314: define noundef range(i64 1, 0) i64 @nonzero_int(i64 noundef range(i64 1, 0) %x) unnamed_addr #1 { 
        315: start: 
        316:  ret i64 %x 
        317: } 
        318:  
        319: ; Function Attrs: uwtable 
        320: define noundef i64 @option_nonzero_int(i64 noundef %x) unnamed_addr #1 { 
check:83                                                           X~~~~~~~~~~~~~~~~~~ error: no match found
        321: start: 
check:83     ~~~~~~~
        322:  ret i64 %x 
check:83     ~~~~~~~~~~~~
        323: } 
check:83     ~~
        324:  
check:83     ~
        325: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        326: define void @readonly_borrow(ptr noalias noundef readonly align 4 dereferenceable(4) %_1) unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        327: start: 
check:83     ~~~~~~~
        328:  ret void 
check:83     ~~~~~~~~~~
        329: } 
check:83     ~~
        330:  
check:83     ~
        331: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        332: define noundef align 4 dereferenceable(4) ptr @readonly_borrow_ret() unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        333: start: 
check:83     ~~~~~~~
        334:  br label %bb1 
check:83     ~~~~~~~~~~~~~~~
        335:  
check:83     ~
        336: bb1: ; preds = %bb1, %start 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        337:  br label %bb1 
check:83     ~~~~~~~~~~~~~~~
        338: } 
check:83     ~~
        339:  
check:83     ~
        340: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        341: define void @static_borrow(ptr noalias noundef readonly align 4 dereferenceable(4) %_1) unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        342: start: 
check:83     ~~~~~~~
        343:  ret void 
check:83     ~~~~~~~~~~
        344: } 
check:83     ~~
        345:  
check:83     ~
        346: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        347: define void @named_borrow(ptr noalias noundef readonly align 4 dereferenceable(4) %_1) unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        348: start: 
check:83     ~~~~~~~
        349:  ret void 
check:83     ~~~~~~~~~~
        350: } 
check:83     ~~
        351:  
check:83     ~
        352: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        353: define void @unsafe_borrow(ptr noundef nonnull align 2 %_1) unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        354: start: 
check:83     ~~~~~~~
        355:  ret void 
check:83     ~~~~~~~~~~
        356: } 
check:83     ~~
        357:  
check:83     ~
        358: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        359: define void @mutable_unsafe_borrow(ptr noalias noundef align 2 dereferenceable(2) %_1) unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        360: start: 
check:83     ~~~~~~~
        361:  ret void 
check:83     ~~~~~~~~~~
        362: } 
check:83     ~~
        363:  
check:83     ~
        364: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        365: define void @mutable_borrow(ptr noalias noundef align 4 dereferenceable(4) %_1) unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        366: start: 
check:83     ~~~~~~~
        367:  ret void 
check:83     ~~~~~~~~~~
        368: } 
check:83     ~~
        369:  
check:83     ~
        370: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        371: define noundef align 4 dereferenceable(4) ptr @mutable_borrow_ret() unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        372: start: 
check:83     ~~~~~~~
        373:  br label %bb1 
check:83     ~~~~~~~~~~~~~~~
        374:  
check:83     ~
        375: bb1: ; preds = %bb1, %start 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        376:  br label %bb1 
check:83     ~~~~~~~~~~~~~~~
        377: } 
check:83     ~~
        378:  
check:83     ~
        379: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        380: define void @mutable_notunpin_borrow(ptr noundef nonnull align 4 %_1) unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        381: start: 
check:83     ~~~~~~~
        382:  ret void 
check:83     ~~~~~~~~~~
        383: } 
check:83     ~~
        384:  
check:83     ~
        385: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        386: define void @notunpin_borrow(ptr noalias noundef readonly align 4 dereferenceable(4) %_1) unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        387: start: 
check:83     ~~~~~~~
        388:  ret void 
check:83     ~~~~~~~~~~
        389: } 
check:83     ~~
        390:  
check:83     ~
        391: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        392: define void @indirect_struct(ptr noalias nocapture noundef readonly align 4 dereferenceable(32) %_1) unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        393: start: 
check:83     ~~~~~~~
        394:  ret void 
check:83     ~~~~~~~~~~
        395: } 
check:83     ~~
        396:  
check:83     ~
        397: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        398: define void @borrowed_struct(ptr noalias noundef readonly align 4 dereferenceable(32) %_1) unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        399: start: 
check:83     ~~~~~~~
        400:  ret void 
check:83     ~~~~~~~~~~
        401: } 
check:83     ~~
        402:  
check:83     ~
        403: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        404: define void @option_borrow(ptr noalias noundef readonly align 4 dereferenceable_or_null(4) %_x) unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        405: start: 
check:83     ~~~~~~~
        406:  ret void 
check:83     ~~~~~~~~~~
        407: } 
check:83     ~~
        408:  
check:83     ~
        409: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        410: define void @option_borrow_mut(ptr noalias noundef align 4 dereferenceable_or_null(4) %_x) unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        411: start: 
check:83     ~~~~~~~
        412:  ret void 
check:83     ~~~~~~~~~~
        413: } 
check:83     ~~
        414:  
check:83     ~
        415: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        416: define void @nonnull_and_nondereferenceable(ptr noundef nonnull %_x) unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        417: start: 
check:83     ~~~~~~~
        418:  ret void 
check:83     ~~~~~~~~~~
```

</details>

wesleywiser reviewed (approved):
Thanks @nikic!

r=me with or without suggestion

nikic commented:
@bors r=wesleywiser

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/rust/actions/runs/17112981238/job/48545890008)
Approved by: wesleywiser
Pushing 125ff8a788c5d6a66917f499abdc00051afe6886 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"wesleywiser","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/rust/actions/runs/17112981238/job/48545890008"},"merge_sha":"125ff8a788c5d6a66917f499abdc00051afe6886"} -->

rust-timer commented:
Finished benchmarking commit (125ff8a788c5d6a66917f499abdc00051afe6886): [comparison URL](https://perf.rust-lang.org/compare.html?start=040a98af70f0a7da03f3d5356531b28a2a7a77e4&end=125ff8a788c5d6a66917f499abdc00051afe6886&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Our benchmarks found a performance regression caused by this PR.
This might be an actual regression, but it can also be just noise.

**Next Steps**:

- If the regression was expected or you think it can be justified,
please write a comment with sufficient written justification, and add
`@rustbot label: +perf-regression-triaged` to it, to mark the regression as triaged.
- If you think that you know of a way to resolve the regression, try to create
a new PR with a fix for the regression.
- If you do not understand the regression or you think that it is just noise,
you can ask the `@rust-lang/wg-compiler-performance` working group for help (members of this group
were already notified of this PR).

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=040a98af70f0a7da03f3d5356531b28a2a7a77e4&end=125ff8a788c5d6a66917f499abdc00051afe6886&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.8%  | [0.8%, 0.8%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 0.3%  | [0.2%, 0.4%]   | 7     |
| Improvements ✅ <br /> (primary)   | -0.2% | [-0.2%, -0.2%] | 1     |
| Improvements ✅ <br /> (secondary) | -1.3% | [-1.7%, -0.2%] | 8     |
| All ❌✅ (primary)                 | 0.3%  | [-0.2%, 0.8%]  | 2     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=040a98af70f0a7da03f3d5356531b28a2a7a77e4&end=125ff8a788c5d6a66917f499abdc00051afe6886&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=040a98af70f0a7da03f3d5356531b28a2a7a77e4&end=125ff8a788c5d6a66917f499abdc00051afe6886&stat=cycles:u)
<details>
<summary>Results (secondary -1.7%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 3.7%  | [3.7%, 3.7%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.0% | [-3.3%, -2.6%] | 4     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=040a98af70f0a7da03f3d5356531b28a2a7a77e4&end=125ff8a788c5d6a66917f499abdc00051afe6886&stat=size:linked_artifact)
<details>
<summary>Results (secondary 0.0%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.0% | [0.0%, 0.0%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

**Bootstrap**: 471.047s -> 471.807s (0.16%)
**Artifact size**: 378.31 MiB -> 378.19 MiB (-0.03%)
